File: src/server/plugins/PluginManager.ts
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';
import { spawn } from 'child_process';

export interface PluginManifest {
	name: string;
	description: string;
	version: string;
	command: string;
	args: string[];
	required_env: string[];
}

export class PluginManager {
	private pluginsDir: string;
	private manifests: Map<string, PluginManifest>;

	constructor(dir?: string) {
		this.pluginsDir = dir || path.join(os.homedir(), '.maestro', 'plugins');
		this.manifests = new Map();

		if (!fs.existsSync(this.pluginsDir)) {
			try {
				fs.mkdirSync(this.pluginsDir, { recursive: true });
			} catch (e: any) {
				console.error(
					`Warning: Failed to create plugin directory ${this.pluginsDir}: ${e.message}`
				);
			}
		}

		this.loadPlugins();
	}

	private loadPlugins() {
		if (!fs.existsSync(this.pluginsDir)) return;

		const entries = fs.readdirSync(this.pluginsDir, { withFileTypes: true });

		for (const entry of entries) {
			if (entry.isDirectory()) {
				const manifestPath = path.join(this.pluginsDir, entry.name, 'manifest.json');
				if (fs.existsSync(manifestPath)) {
					try {
						const data = fs.readFileSync(manifestPath, 'utf8');
						const manifest: PluginManifest = JSON.parse(data);
						this.manifests.set(manifest.name, manifest);
						console.log(`Loaded plugin: ${manifest.name} (v${manifest.version})`);
					} catch (e: any) {
						console.error(`Failed to load plugin manifest at ${manifestPath}: ${e.message}`);
					}
				}
			}
		}
	}

	public getAvailablePlugins(): PluginManifest[] {
		return Array.from(this.manifests.values());
	}

	public async *executePlugin(name: string, input: string): AsyncGenerator<string, void, unknown> {
		const manifest = this.manifests.get(name);

		if (!manifest) {
			throw new Error(`Plugin not found: ${name}`);
		}

		// Mock execution for skeleton
		yield JSON.stringify({
			status: 'streaming',
			data: `Executed plugin: ${name} with input: ${input}`,
		});
	}
}

File: go/internal/plugins/plugin_manager.go
package plugins

import (
	"encoding/json"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"sync"
)

type PluginEvents struct {
	OnStart       string `json:"on_start,omitempty"`
	OnFileChanged string `json:"on_file_changed,omitempty"`
	OnError       string `json:"on_error,omitempty"`
}

type PluginManifest struct {
	Name        string       `json:"name"`
	Description string       `json:"description"`
	Version     string       `json:"version"`
	Command     string       `json:"command"`
	Args        []string     `json:"args"`
	RequiredEnv []string     `json:"required_env"`
	Events      PluginEvents `json:"events,omitempty"`
}

type PluginManager struct {
	pluginsDir string
	manifests  map[string]PluginManifest
	mu         sync.RWMutex
}

func NewPluginManager(dir string) *PluginManager {
	if dir == "" {
		homeDir, _ := os.UserHomeDir()
		dir = filepath.Join(homeDir, ".maestro", "plugins")
	}

	// Ensure directory exists
	if err := os.MkdirAll(dir, 0755); err != nil {
		log.Printf("Warning: Failed to create plugin directory %s: %v", dir, err)
	}

	pm := &PluginManager{
		pluginsDir: dir,
		manifests:  make(map[string]PluginManifest),
	}
	pm.LoadPlugins()
	return pm
}

func (pm *PluginManager) LoadPlugins() {
	pm.mu.Lock()
	defer pm.mu.Unlock()

	entries, err := os.ReadDir(pm.pluginsDir)
	if err != nil {
		log.Printf("Error reading plugins directory: %v", err)
		return
	}

	for _, entry := range entries {
		if entry.IsDir() {
			manifestPath := filepath.Join(pm.pluginsDir, entry.Name(), "manifest.json")
			if data, err := os.ReadFile(manifestPath); err == nil {
				var manifest PluginManifest
				if err := json.Unmarshal(data, &manifest); err == nil {
					pm.manifests[manifest.Name] = manifest
					log.Printf("Loaded plugin: %s (v%s)", manifest.Name, manifest.Version)
				}
			}
		}
	}
}

func (pm *PluginManager) GetAvailablePlugins() []PluginManifest {
	pm.mu.RLock()
	defer pm.mu.RUnlock()

	var plugins []PluginManifest
	for _, p := range pm.manifests {
		plugins = append(plugins, p)
	}
	return plugins
}

func (pm *PluginManager) EmitEvent(eventName string, payload interface{}) {
	pm.mu.RLock()
	defer pm.mu.RUnlock()

	for _, manifest := range pm.manifests {
		commandToRun := ""
		switch eventName {
		case "on_start":
			commandToRun = manifest.Events.OnStart
		case "on_file_changed":
			commandToRun = manifest.Events.OnFileChanged
		case "on_error":
			commandToRun = manifest.Events.OnError
		}

		if commandToRun != "" {
			log.Printf("[PluginManager] Emitting %s to %s via %s", eventName, manifest.Name, commandToRun)
			// Mock exec
		}
	}
}

func (pm *PluginManager) ExecutePlugin(name string, input string) (string, error) {
	pm.mu.RLock()
	manifest, exists := pm.manifests[name]
	pm.mu.RUnlock()

	if !exists {
		return "", os.ErrNotExist
	}

	// Create command
	cmd := exec.Command(manifest.Command, manifest.Args...)
	_ = cmd // ignore for mock
	// In a real implementation, you would pipe input via stdin,
	// set environment variables from the ConfigManager,
	// and read stdout iteratively to stream results back.

	// For this skeleton, we just mock the return
	return "Executed plugin: " + name + " with input: " + input, nil
}
